### ISSUE TITLE
Qubes-provided kernels are out of date

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4846]
### Qubes OS version:

`Qubes release 4.0 (R4.0)`

### Affected component(s) or functionality:

dom0 and dom0-provided VM kernels

---

### Steps to reproduce the behavior:

Run `uname -r` in dom0

### Expected or desired behavior:

Reports the latest kernel in its branch (4.14, etc)

### Actual behavior:

Kernel is may (30+!) patch releases out of date, and so misses out on upstream bug-fixes.

### General notes:

Upstreaming Qubes’s kernel patches would help.

---

**Comment by marmarek** [comment]
@fepitre is working on a script automating this.

**Comment by marmarek** [comment]
In the meantime, I'll update stable-4.14 branch manually.

**Comment by DemiMarie** [comment]
@fepitre @marmarek any updates?

**Comment by DemiMarie** [comment]
Upstream is now on 4.19.29, I believe.

**Comment by marmarek** [comment]
And there is 4.19.29 in current-testing repository. But this is still done manually...

**Comment by fepitre** [comment]
@DemiMarie: I'm still working on auto updates besides others things...in the next weeks it should be done.

**Comment by fepitre** [comment]
I think we can close this issue, if not please feel free to reopen it. The automatic process is on since several weeks for updating all the stable and master branches with https://github.com/fepitre/qubes-linux-kernel-updater/.

**Comment by DemiMarie** [comment]
@fepitre @andrewdavidwong Can this be reopened?  The `kernel-latest` package is still several versions behind.  I could also file a separate issue.

**Comment by fepitre** [comment]
Latest stable is 5.2.10. We have kernel-latest-5.2.9-1.pvops.qubes.x86_64.rpm in repo. The automatic build every Saturday morning but yesterday kernel 5.2.10 was not out. In that case, what do you mean also by several versions behind?

**Comment by hexagonrecursion** [comment]
fepitre, you are right: there is kernel-latest-5.2.9, but it is in the qubes-dom0-current-testing repository. DemiMarie is probably referring to the fact that the newest kernel-latest is qubes-dom0-current is kernel-latest-5.1.15.

**Comment by fepitre** [comment]
@hexagonrecursion: yes indeed. I think @marmarek does not want to push quickly these latest kernels into 'current' due to high recurrent troubles with newer kernels/hardware. @marmarek can you confirm that? Also, I suggested once to do like Fedora, a testing campaign for newer kernels but that would require a lot of users and time but I'm in for that :)

**Comment by marmarek** [comment]
I'm not sure about the need for a written policy here. 

What we do need to improve the situation is people testing packages in current-testing repo **and providing feedback (including positive one)**, as documented here: https://www.qubes-os.org/doc/testing/#providing-feedback

**Comment by marmarek** [comment]
> whereas @fepitre (and maybe @marmarek) do not expect such new kernels to be in the `current` repo, at least not until they have been sufficiently tested, which usually takes some time.

Yes, every update first is uploaded to "current-testing" (and to "security-testing" if that's a security fix). It is later added to "current" repository after it get sufficiently tested, but not earlier than a week after current-testing upload. "Sufficiently tested" in practice is fluid term, and means no negative feedback plus at least one positive feedback or staying in current-testing long enough (subjective, depending on component, complexity of changes, etc).

There is also a limitation of our current tooling: it is possible to move current-testing->current only the most recent version of the package. This means if newer version of a package is uploaded to current-testing, it isn't possible to move older one to current, even if it was there long enough and is deemed stable. While in some cases it may be inconvenient (especially given enforced week in current-testing + weekly kernel builds), generally I think it eases testing and reporting (always talk about latest curren-testing version, not some arbitrary older one).

Anyway, I see very few people providing feedback on testing packages in https://github.com/QubesOS/updates-status/issues. I guess too few people know about it... But also maybe it's too inconvenient? Maybe we need a tool to automate some of it at least ("report-my-current-testing-system-works-ok" tool)?

**Comment by marmarek** [comment]
> > I expect `current-testing` to be the latest release from `kernel.org` at the time it was built.
> 
> It sounds like this is the main point of conflicting expectations. @marmarek, thoughts?

I don't see any conflict here. It is true, as explained in https://github.com/QubesOS/qubes-issues/issues/4846#issuecomment-524662220

**Comment by marmarek** [comment]
'kernel' is LTS kernel, more tested. 'kernel-latest' is latest (possibly non-LTS) version, less tested. See https://www.qubes-os.org/doc/newer-hardware-troubleshooting/

**Comment by fepitre** [comment]
We discussed once with @marmarek to possibly add a testing pipeline to openqa for kernel-latest. At least, that would be validate or not all the current Qubes tests with newer kernels but in an virtual environment. Any thought about that?

**Comment by HW42** [comment]
>> Anyway, I see very few people providing feedback on testing packages in https://github.com/QubesOS/updates-status/issues. I guess too few people know about it... But also maybe it's too inconvenient? Maybe we need a tool to automate some of it at least ("report-my-current-testing-system-works-ok" tool)?
>
> To be honest, I don't think that most people will exert any effort until they have a problem, and even then, they'll exert the minimum effort required to get their problem solved (or just give up). So, yes, if we need more data, we should set up some kind of automated (privacy-respecting, user-consent-required) reporting tool.

That's probably true for the majority. But if I look at the latest packages in updates-status nobody (except @DemiMarie, after this discussion) reports positive feedback there. So it might be worth a try to remind the bunch of more active people that this is wanted. Even 1..2 reports from others help.

> We discussed once with @marmarek to possibly add a testing pipeline to openqa for kernel-latest. At least, that would be validate or not all the current Qubes tests with newer kernels but in an virtual environment. Any thought about that?

That surely doesn't hurt. But I think for the kernel you want to also test on real hardware, because especially things like suspend tend to break.

> But also maybe it's too inconvenient? Maybe we need a tool to automate some of it at least ("report-my-current-testing-system-works-ok" tool)?

I think this would help, but is tricky to implement. You would need to track updated packages and then maybe 1..2 days later (but only if they rebooted into the new kernel ...) ask the users if everything looks fine.

Another part of the problem is probably that having a test system which you don't only use if you really need to test something is quite some extra effort, and on your main system many probably want to be a bit more conservative with using testing versions.

**Comment by andrewdavidwong** [comment]
> So it might be worth a try to remind the bunch of more active people that this is wanted. Even 1..2 reports from others help.

Ok, I've sent a reminder to the mailing lists:

https://groups.google.com/d/topic/qubes-devel/8tyWbAhpFSM/discussion
https://groups.google.com/d/topic/qubes-users/3c5vXv-eRjo/discussion

**Comment by caissalover** [comment]
> I'm not sure about the need for a written policy here.
> 
> What we do need to improve the situation is people testing packages in current-testing repo **and providing feedback (including positive one)**, as documented here: https://www.qubes-os.org/doc/testing/#providing-feedback

Using meant to be stable kernels, we're facing problems like not being able to boot up and use networks. I think once they become sufficiently stable for absolutely basic problems, people would surely try latest kernels. Now, I'm still afraid about the stable ones.

**Comment by marmarek** [comment]
During last month+ we've got few users giving positive feedback (thanks @hexagonrecursion @awokd @strugee @0spinboson @DemiMarie ), which allowed faster upload to stable.
